File: packages/jekko/src/cli/cmd/tui/feature-plugins/jankurai/sparkline.ts
// Pure unicode-block sparkline renderer. Used by the audit-live panel to show
// score history in a single line of width N. Width-clip from the tail so the
// most-recent samples always appear; resolution maps (max - min) onto eight
// glyphs.

const GLYPHS = ["▁", "▂", "▃", "▄", "▅", "▆", "▇", "█"] as const
const PLACEHOLDER = "·"

export function sparkline(values: readonly number[], width: number): string {
  if (width <= 0) return ""
  if (values.length === 0) return PLACEHOLDER.repeat(width)
  const tail = values.length > width ? values.slice(values.length - width) : values
  // Min / max over finite values only — otherwise a NaN poisons the bounds
  // and every glyph collapses to a placeholder.
  const finite = tail.filter((v) => Number.isFinite(v))
  if (finite.length === 0) {
    return PLACEHOLDER.repeat(width)
  }
  const min = Math.min(...finite)
  const max = Math.max(...finite)
  const span = max - min
  const glyphsForTail = tail.map((value) => {
    if (!Number.isFinite(value)) return PLACEHOLDER
    if (span === 0) return GLYPHS[Math.floor(GLYPHS.length / 2)]
    const idx = Math.min(
      GLYPHS.length - 1,
      Math.max(0, Math.floor(((value - min) / span) * (GLYPHS.length - 1) + 0.5)),
    )
    return GLYPHS[idx]
  })
  if (glyphsForTail.length >= width) return glyphsForTail.join("")
  // Left-pad with placeholder so the sparkline lines up at the right edge.
  return PLACEHOLDER.repeat(width - glyphsForTail.length) + glyphsForTail.join("")
}

export const SPARKLINE_GLYPHS = GLYPHS
export const SPARKLINE_PLACEHOLDER = PLACEHOLDER

File: crates/jekko-tui/src/feature_plugins/jankurai/sparkline.rs
//! Unicode-block sparkline renderer.

pub(super) const GLYPHS: &[char] = &['▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'];
pub(super) const BLANK_GLYPH: char = '·';

/// Unicode-block sparkline renderer. Ports `sparkline.ts::sparkline`.
///
/// Tails to the right so the most recent samples always appear; pads with
/// blank glyphs when there are fewer samples than `width`.
pub fn sparkline(values: &[f64], width: usize) -> String {
    if width == 0 {
        return String::new();
    }
    if values.is_empty() {
        return BLANK_GLYPH.to_string().repeat(width);
    }
    let tail: Vec<f64> = if values.len() > width {
        values[values.len() - width..].to_vec()
    } else {
        values.to_vec()
    };
    let finite: Vec<f64> = tail.iter().copied().filter(|v| v.is_finite()).collect();
    if finite.is_empty() {
        return BLANK_GLYPH.to_string().repeat(width);
    }
    let mut min = finite[0];
    let mut max = finite[0];
    for v in &finite {
        if *v < min {
            min = *v;
        }
        if *v > max {
            max = *v;
        }
    }
    let span = max - min;
    let mut out: String = tail
        .iter()
        .map(|v| {
            if !v.is_finite() {
                return BLANK_GLYPH;
            }
            if span == 0.0 {
                return GLYPHS[GLYPHS.len() / 2];
            }
            let frac = (v - min) / span;
            let idx = (frac * (GLYPHS.len() - 1) as f64 + 0.5).floor() as usize;
            let clamped = idx.min(GLYPHS.len() - 1);
            GLYPHS[clamped]
        })
        .collect();
    if out.chars().count() < width {
        let pad = width - out.chars().count();
        let mut prefix = String::new();
        for _ in 0..pad {
            prefix.push(BLANK_GLYPH);
        }
        prefix.push_str(&out);
        out = prefix;
    }
    out
}
